LiveSearch also searches in Status + displays results count

soif opened the pull request:
as asked in #5518

enjoy 😎

soif committed:
adds data-status parameter to device items

soif committed:
LiveSearch: Allow to also search by Status

soif committed:
LiveSearch Input's Placeholder : adds Idx and Status

gizmocuz commented on www/js/domoticz.js:
```diff
@@ -6317,13 +6317,14 @@ function WatchLiveSearch(){
 				var name=$(this).find('#name').html().toUpperCase();
 				var desc=$(this).find('#name').attr('data-desc');
 				var idx=$(this).find('#name').attr('data-idx');
+				var status=$(this).find('#name').attr('data-status').toUpperCase();
```
Just checking, if (for example for a utility devices) there is no status, would this not give an assertion?
Maybe status should be checked like 'desc' below, if it is undefined, we make it empty else we make it uppercase?

soif replied:
my bad! You're right!

_--> I always forget that there is currently no global template for devices 😉_ 

A fix will follow ASAP.

soif replied:
fixed!

gizmocuz commented:
Fast as lightning ;)

soif committed:
LiveSearch: Defaults undefined data attributes to empty string

soif committed:
LiveSearch: use Regex to search (better language support)+ 'liveSearchShown' & 'itemBlock' as variables

soif committed:
Comments

soif commented:
- I fixed the `undefined` case
- Search now uses RegEx: It should be better for internationalization + allow typing some basic Regex, ie `^name`  (starts with) or `pattern$` (ends with)
- Status are searched only from the **start** of the data-status txt (better filtering IMHO)

soif committed:
LiveSearch: Displays results count with a Close/reset icon

soif committed:
LiveSearch: Style for Search results count

soif commented:
As a Bonus:
- Search results count is displayed while searching.
- A close icon allows to quickly reset the search

Hope you like it?

soif committed:
Merge branch 'development' into LiveSearchStatus

kiddigital reviewed (commented):
@soif , it is good practice to keep a PR branch up-to-date with its base branch to prevent any merge conflicts when the PR gets merged. (so thx for that).

As an alternative to `merge` the updated base branch back into the PR branch (which creates a merge-commit in the PR branch), you can _rebase_ the PR branch on its base branch (`git rebase -i upstream/development` for example). This effectively takes the updated base branch and tries to re-apply all commits in the PR branch.

Great thing also is that in case of any merge conflicts, they will appear during the rebase and you can fix them immediately. The necessary changes to fix the merge conflicts become part of the commit that caused the merge conflict, so no extra commit(s) needed to get the intended result of the initial commit but now applied on the updated base branch. Than you can `git push -f` the PR branch again and the PR branch is up-to-date and ready to be merged.

This way you keep the PR branch both up-to-date and ready for merging without having extra merge commits and possible merge conflict solving commits :)

soif commented:
@kiddigital 
Thank for the helpful explanation. 😉 
I'm aware of the git rebase mechanism, but I've never used it.
_I did not found the time to make a test project, and then experiment how the rebase works... But I promise I will do it one day or later..._ 😎

gizmocuz commented:
Thank you for this PR!! Looking good... two small issues

When I search for 'on' it also displays 'off", same for 'off' it displays also the 'on' switches

![image](https://user-images.githubusercontent.com/2265299/211542079-1c91291a-609d-4711-bea2-49ddb41a4d7a.png)

This is probably caused by the type indication? Maybe we have to change the on/off to normal, or ignore this on/off?

The font with the counter could maybe be a little bolder or bigger... it's a bit small

soif commented:
>When I search for 'on' it also displays 'off", same for 'off' it displays also the 'on' switches

I dont know for the **Off** case, but regarding the **On** case,  from your screenshot,  the device is shown because the name is  testbutt**ON** 😛 

>This is probably caused by the type indication? 

Nope, type is not searched, only fields in the data-XXX parameters...

>The font with the counter could maybe be a little bolder or bigger... it's a bit small

I've done this because while the count info is interesting, its not VERY important, so it can be smaller.

I could make it a bit larger, but as you may have noticed, ASAP as the count is displayed, the search field width is reduced (by the average amount of the count field width) so that, the following blocks (ie Room plans) are not shifted to the right (even more important on mobiles, which could lead to have plans popup to "jump "to the next line.)

So I can enlarge the font, but it will reduce the search field a little bit more.. Do U Want it?

gizmocuz commented:
@soif , you're right ;) it's because 'on' is part of the name , same on a device that has 'motion' in the name ;)
No please leave the count as it is ;)

Thinking further, I'm on my Utility view and wanted to quickly see all my kWh sensors, or Usage/Watt sensors, or Lux sensors, or percentage sensors, or all my custom sensors...
Maybe for this view it would be nice to have the 'type' field searchable?

soif committed:
LiveSearch: also search in Status from the Utility page

soif committed:
No longer search Status from the start (needed for the Utility page)

soif commented:
Per request, I just added Search in the Status from the Utility page.

I rather search in "Status", than  in "Type", because:
- searching the unit used in the Status (ie "Watt", "° C", "Kb/s") mostly archieves the same  results
- It make the Search UI more consistent: The user expects to search in "Name,Desc,Ids,Status" everywhere, but not in the "type"

Note that to allow to search _somewhere_ in the Status, I've removed the RegEx that was searching only from start.

Sounds Good?

soif committed:
LiveSearch: also search in Status from the Scenes page

soif commented:
New Bonus: Also searches Status from the Scenes page.

_--> again this makes the search UI more consistent_

------

BTW: I've also intended to add Status Search in Temperatures/wheather pages, BUT it is very very very tedious, because of a design flaw (IMHO) in these pages.. 

_--> The final Status (=BigText) is built from the Angular template (xxx_widget.html), rather than from the controller._

 Might I suggest that when you'd redesign the whole Angular framework, you might add a global JS function ie _"LoadItemsFormatted(url)"_ (that would be called by each page controller), that loops thru each retrieved item, and formats each item (maybe in a sub _FormatItem(item) function), to create a new **item.html_value** (=BigText=Status) in the _items_ object. That way, each xxx_widget.html template just have to print the **item.html_value** without having to worry about complex types combinations logic...

_This item.html_value, could also be directly built from the C code, so that it would also appears in API answers..._

 See what I mean?
 Makes sense ?

gizmocuz commented:
Thank you once again! Your fast!!! 😁
Yes it makes sense to refactor the HTML/JS code, but to be honest I think it will take quite some time to make a new UI based on typescript/Angular instead of AngularJS.
But typescript makes it a lot more error free

About the generic function (js or in c++) that's a good idea. In principle the idea was to use the 'Data' field.
But some sensors have multiple values (like a kWh sensor (today counter and current usage, same for P1 meter, gas and a few others)
Maybe for some devices the 'Data' field should be extended with the unit
But for most of the devices the bigtext='Data' field

soif commented:
> Thank you once again! Your fast!!! 😁 

I don't like to leave a PR not **fully** finished... 😇 

>Yes it makes sense to refactor the HTML/JS code, but to be honest I think it will take quite some time to make a new UI 

Huge work, I know.... 😲 

> About the generic function (js or in c++) that's a good idea. In principle the idea was to use the 'Data' field.

IMO it would be best in the C++ code!!!
-  API users wouldn't have to fully replicate this logic in their "**Views**", anymore to correctly print data, whatever the type! 
- That's the way a smart API is usually built... 😺
- On the other hand, it would allow to drastically simplify the actual JS/Angular code ASAP, without having to wait for the complete Angular refactoring.... 

------

ready/happy to merge ?

gizmocuz commented:
I have tested your additions ;) ... I still have a problem finding my kWh devices in the utility tab
In the seach box I enter 'kWh' and I only see the devices that have this text inside the title/name

But this device is not found

![image](https://user-images.githubusercontent.com/2265299/211816591-d7057cbd-0ae7-484f-9949-f0b2b7e6e48a.png)

gizmocuz commented:
The text kWh is also in the today, but also in the 'type'

soif commented:
>I rather search in "Status", than in "Type", because:

Again 😉 only STATUS (aka BigText) is searched, not TYPE...

Search for "Watt".... 🙄

gizmocuz commented:
Yes, that will work but, on some systems, this produces 50% wrong/not needed results 😁
If you have a additional data field to search for (which is never used, but in this case filled with the Type for Utility sensors) this would do the trick
At the moment you have multiple data types, maybe it is an idea to have one and add all data into this you want to search for?
It would maybe make the search function simpler and faster?
But happy to merge if you say it's ready to be!

soif commented:
> Yes, that will work but, on some systems, this produces 50% wrong/not needed results 😁 

That's obviouly the drawback of searching MULTIPLE fields... I can not do any magic here ATM 🙄 
_Any better idea is welcome !_

>If you have a additional data field to search for (which is never used, but in this case filled with the Type for Utility sensors) this would do the trick 

I can, but you initially asked to search in the "Status" field, and this is what this PR basically does...

>At the moment you have multiple data types, maybe it is an idea to have one and add all data into this you want to search for? 

Yep, I might add **type** search (and certainly also subtype search) to some (not all) pages... But can we add/discuss this in another PR?

>It would maybe make the search function simpler and faster? 

Nope (or I did not understood). It would simply add another field searched, in some case (ie :only in the Utility page).... This would make the Search UI less consistent, because only, in this case (Utility page), the search would ALSO search in type...

>But happy to merge if you say it's ready to be!

NO, you're the master of merge, and only YOU can say if it deserves to be merged ATM, or not 😄 

From my point of view, this PR fits your initial request, but  if you want me to first improve it b4 merging, I'm here to serve!

gizmocuz commented:
Thanks again!!

soif commented:
Thanks for the merge! 

You're also (with @kiddigital ) nominated as the the fastest merger of the universe... 🤣  🤣  🤣 

If needed, please feel free to start an issue (notifying me) to also search in Type, in some cases...

Best regards

gizmocuz commented:
Thanks, Okey I will spawn an idea in a new issue ;)
